targets/TARGET_STM/TARGET_STM32F7/device/stm32f7xx_hal_cryp.c

```diff
@@ -713,4456 +702,5523 @@ __weak void HAL_CRYP_MspDeInit(CRYP_HandleTypeDef *hcryp)
+        hcryp->Instance->K2RR = *(uint32_t*)(hcryp->Init.pKey+3);
+        hcryp->Instance->K3LR = *(uint32_t*)(hcryp->Init.pKey+4);
+        hcryp->Instance->K3RR = *(uint32_t*)(hcryp->Init.pKey+5);  
       }
-    }
+      
+      /*Set Initialization Vector (IV)*/    
+      if ((hcryp->Init.Algorithm == CRYP_DES_CBC) || (hcryp->Init.Algorithm == CRYP_TDES_CBC))
+      {
+        hcryp->Instance->IV0LR = *(uint32_t*)(hcryp->Init.pInitVect);
+        hcryp->Instance->IV0RR = *(uint32_t*)(hcryp->Init.pInitVect+1);
+      }
+      
+      /* Flush FIFO */
+      HAL_CRYP_FIFO_FLUSH(hcryp);
+      
+      /* Set the phase */
+      hcryp->Phase = CRYP_PHASE_PROCESS;
+      
+      /* Statrt DES/TDES encryption process */     
+      status = CRYP_TDES_Process(hcryp,Timeout);     
+      break;
+      
+    case CRYP_AES_ECB:
+    case CRYP_AES_CBC:
+    case CRYP_AES_CTR:
+      
+      /* AES encryption */       
+      status = CRYP_AES_Encrypt(hcryp, Timeout);
+      break;   
+      
+    case CRYP_AES_GCM:
+      
+      /* AES GCM encryption */       
+      status = CRYP_AESGCM_Process(hcryp, Timeout);
+      
+      break;
+      
+    case CRYP_AES_CCM:
+      
+      /* AES CCM encryption */  
+      status = CRYP_AESCCM_Process(hcryp,Timeout);     
+      break;
+       
+    default:
+      hcryp->ErrorCode |= HAL_CRYP_ERROR_NOT_SUPPORTED;
+      return HAL_ERROR;
+    }      
     
-    /* Disable CRYP */
-    __HAL_CRYP_DISABLE(hcryp);
+#else /*AES*/
     
-    /* Reset the ALGOMODE bits*/
-    CRYP->CR &= (uint32_t)(~CRYP_CR_ALGOMODE);
+    /* Set the operating mode*/
+    MODIFY_REG(hcryp->Instance->CR, AES_CR_MODE, CRYP_OPERATINGMODE_ENCRYPT); 
     
-    /* Set the CRYP peripheral in AES ECB decryption mode */
-    __HAL_CRYP_SET_MODE(hcryp, CRYP_CR_ALGOMODE_AES_ECB | CRYP_CR_ALGODIR);
-    /* Flush FIFO */
-    __HAL_CRYP_FIFO_FLUSH(hcryp);
+    /* algo get algorithm selected */  
+    algo = hcryp->Instance->CR & AES_CR_CHMOD;  
     
-    /* Enable CRYP */
-    __HAL_CRYP_ENABLE(hcryp);
+    switch(algo)
+    {    
+      
+    case CRYP_AES_ECB:
+    case CRYP_AES_CBC:
+    case CRYP_AES_CTR:
+      
+      /* AES encryption */        
+      status = CRYP_AES_Encrypt(hcryp, Timeout);     
+      break;   
+      
+    case CRYP_AES_GCM_GMAC:
+      
+      /* AES GCM encryption */  
+      status = CRYP_AESGCM_Process (hcryp,Timeout) ;  
+      break;
+      
+    case CRYP_AES_CCM:
+      
+      /* AES CCM encryption */ 
+      status = CRYP_AESCCM_Process(hcryp,Timeout);   
+      break;
+      
+    default:
+      hcryp->ErrorCode |= HAL_CRYP_ERROR_NOT_SUPPORTED;
+      return HAL_ERROR;
+    }
+#endif /*end AES or CRYP */   
     
-    /* Set the phase */
-    hcryp->Phase = HAL_CRYP_PHASE_PROCESS;
+    if (status == HAL_OK)
+    { 
+      /* Change the CRYP peripheral state */
+      hcryp->State = HAL_CRYP_STATE_READY;
+      
+      /* Process unlocked */
+      __HAL_UNLOCK(hcryp);
+    }    
```

Comment:
Do we need to `__HAL_UNLOCK(hcryp);` when `status != HAL_OK`? If not, why not?

Same goes `HAL_CRYP_AESCTR_Decrypt()` and other similar functions.

Reply:
ST_INTERNAL_REF 75115